### ISSUE TITLE
Nested use of produce?

### ISSUE THREAD

**Issue by ianstormtaylor** [immerjs/immer#433]
Hello, I've tried searching through the existing issues and reading through the documentation, but I'm still not exactly clear on how nested `produce` calls are designed to work. Could this be added to the docs?

It seems like a common use case to me, with a simple example being:

```js
const increment = (counter) => {
  return produce(counter, c => {
    c.count++
  })
}
```

If somewhere else I want to reuse the `increment` logic, can I do:

```js
produce(counter, c => {
  increment(c)
})
```

Does that work seamlessly? With the nested call to `produce` seeing that `c` is already a draft, and just passing through to let it be modified? Thanks!

**Comment by mweststrate** [comment]
Yes that should work seamlessly and 'draftness' should be preserved when
passing a draft in as base state to a producer

**Comment by ianstormtaylor** [comment]
Beautiful, thank you!

**Comment by ianstormtaylor** [comment]
Actually, that begs a follow up which I think should be documented too: does that apply for all nested objects inside the initial draft? Or does the "draft" only refer to the top-level object?

For example:

```js
const increment = (counter) => {
  return produce(counter, c => {
    c.count++
  })
}

const incrementAll = (counters) => {
  return produce(counters, cs => {
    for (const c of cs) {
      increment(c)
    }
  })
}
```

Is it right to say that in all cases: 

Any function which uses `produce` to return a new value could be called a "producer". In normal execution, producers don't mutate their inputs, and instead return a new immutable value. **However**, when producers are called inside of a `produce()` callback, they **do** mutate their inputs.

**Comment by mweststrate** [comment]
Yes :). But just trying it in a sandbox is a quicker way of finding that
out ;-). Feel free to PR the docs to clarify!

**Comment by ianstormtaylor** [comment]
Thanks. A sandbox can work! But I'm more interested in what the mental model is and how to structure things, than what happens to work in whatever test case I write up.

With functions like `increment`, the above behavior makes sense.

But... it seems like the current design—where a `produce()` in the call stack changes the mutability of producers further down the stack—limits `immer` to only being used for transform-ish functions? And that it can't really be used as internals for writing cleaner get-ish functions?

For example, consider:

```js
const getParent = path => {
  return produce(path, p => {
    p.pop()
  })
}

const getNext = path => {
  return produce(path, p => {
    p[p.length - 1]++
  })
}
```

In this case, I'm just using `produce` for convenience for the internals of a these functions that need to ensure they act without mutating.

But this seems like this creates an unexpected problem. If these functions happen to be called inside a `produce()` higher up the stack and the `path` argument was some nested object in that draft state, they no longer just get a related path, they now mutate the input path itself?

Is that right?

**Comment by mweststrate** [comment]
could you provide real and running examples? I find your examples really
hard to follow; you have getter functions with side effects? That
sounds terribly confusing in any case, regardless whether you use immer or
not. If you don't intend to have side effects, just create a utility like
`peek`, instead of using `pop`. That will be much clearer for everybody
involved.

**Comment by ianstormtaylor** [comment]
Absolutely, happy to elaborate. I think the reason you don't understand my examples is because we have slightly different mental models for what we think immer does—and "side effects" is a great way to put it. (So I ask that you give me the benefit of the doubt a little bit, I'm not trying to annoy you.)

As someone just coming to immer, after reading the docs...

> Immer (German for: always) is a tiny package that allows you to work with immutable state in a more convenient way.

> This means that you can interact with your data by simply modifying it while keeping all the benefits of immutable data.

> The interesting thing about Immer is that the `baseState` will be untouched, but the `nextState` will be a new immutable tree that reflects all changes made to `draftState` (and structurally sharing the things that weren't changed).

> …

There are actually two different mental models you can take away from these. They are only very slightly different, but they have a big impact on whether you run into unexpected behaviors:

1. Immer allows you to write **side-effects** easily, by expressing them as mutable instructions to make on an initial state.

2. Immer allows you to write immutable functions more conveniently, by letting you write mutable instructions but produce immutable values. 

Especially if you're someone coming from Immutable.js, the 2nd mental model is going to feel familiar, because it's very similar to how you might think about the `withMutations` helper.

So... if you apply these to code...

(Again, please... benefit of the doubt here. These examples are very simplified, but it should be possible to imagine that the mutating logic in either of them could be much more complex. Either a complex side-effect or a complex getter computation.)

```js
// Model #1: You'd only write your "side-effect" functions using immer
// and your non-side-effect functions using traditional immutable ways.
const increment = path => {
  return produce(path, p => {
    p[p.length - 1]++
  })
}

const getParent = path => {
  return path.slice(0, -1)
}
```
```js
// Model #2: You're not limited to just side-effect functions. You use
// immer as convenient for mutating to arrive at an immutable value.
const increment = path => {
  return produce(path, p => {
    p[p.length - 1]++
  })
}

const getParent = path => {
  return produce(path, p => {
    p.pop()
  })
}
```

In case it's not obvious, the 2nd model is very similar to what you could do with Immutable.js's `withMutations` helper as well: (Not that I'm saying they are equivalent!)

```js
// Immutable.js
const increment = path => {
  return path.withMutations(p => {
    const i = p.size - 1
    p.set(i, p.get(i)++)
  })
}

const getParent = path => {
  return path.withMutations(p => {
    p.pop()
  })
}
```

But with the current nested `produce()` behavior, the 2nd model actually leads to unexpected bugs, only when the `getParent` method is wrapped in a `produce()` call.

This is made even more clear if instead of `getParent`, we implemented `getNextSibling`:

```js
// Model #2
const getNextSibling = path => {
  return increment(path)
}
```

It would be tempting to use `increment()` here as a shorthand, since it returns an immutable value that is equivalent to the next sibling. And if as the end user your mental model for the increment helper is that it just returns a new immutable path, you wouldn't think twice about it. But again this has the same bug.

---

Which is where the "side effects" distinction comes in. 

If you truly think of immer as "applying side effects", then you might be able to avoid this bugs. But I think it's really counter-intuitive to say that a function "has side effects" when using immer, since the whole point is eliminating side effects by returning with immutable state... (And I'd assume I'm not alone in thinking that.)

To me it seems like the nested `produce()` behavior introduces a significant amount of emergent complexity into the system. It seems like a footgun. You have to always be aware of the callstack you're in, which as an exported utility function in a library is impossible.

For React-specific Redux-like use cases I think this might not ever crop up, because you're only ever processing a single reducer at once. But if you're trying to use `immer` for a framework or set of utility functions, to replace Immutable.js, then you get tripped up?

Does that make sense?

**Comment by ianstormtaylor** [comment]
For a potential solution... Right now the issue is that the most common way people use immer is via the `produce` singleton:

```js
import { produce } from 'immer'
```

And that singleton is configured to have the problematic cascading mutations behavior.

If instead, the cascading behavior was configurable, opt-in and defaulted to `false`... that `produce` singleton would be safe by default. You could use it anywhere in any codebase, without fear that your logic could be broken at any time by a `produce()` call higher up the callstack that you had no control over.

Then for the cases that really do benefit from the higher performance of cascading draft mutations, they could opt in to that behavior by initializing their own instance of immer with custom configuration that enabled that feature.

Seems like that could unlock immer for a wider range of use cases? While also eliminating the potential footgun that exists right now that many might not be aware of?

**Comment by mweststrate** [comment]
The goal of immer is to be able to write functions that mutate the draft.

That doesn't seem to be the goal of the functions you are writing (rather
the opposite, you want to loose those and keep some derived data), so I
think it is simpler to say Immer isn't designed for your case.

Iirc if you create a separate Immer instance, you get a separate produce
function, that doesn't share drafts with the default instance, so that
might work for you.

Personally I think that your examples are still unnecessarily confusing
though by using side effectfull APIs where you don't intent them.

Op za 5 okt. 2019 04:09 schreef Ian Storm Taylor <notifications@github.com>:

**Comment by ianstormtaylor** [comment]
I don’t think you’ve understood what I’m trying to say. 

Do you see how it’s counter intuitive that immer—a library concerned with producing immutable values and allow you to guarantee immutability—has to concern itself with whether your functions “have side-effects”? And not even *real* side effects, but “if this function was really mutable, would it have had side effects?”  Despite the entire point of using immutable structures being to avoid side effects.

You’re absolutely right about initializing my own instance. But what I’m pointing out is two things:

1. That the default configuration leaves open an unexpected surface for bugs, which currently isn’t well documented. Because nesting two calls to the global `produce` unintuitive-ly changes the behavior of code that you haven’t written. 

2. And that the default configuration—in search of better performance—is actually limiting the utility immer could provide by artificially restricting to a single use case. When there’s an adjacent use case that could be equally solvable by eliminating the footgun. 

The thing is... by switching the configuration default you’d actually end up with a safer library **and** one with more utility.

Under the current architecture, using nested `produce` calls requires you to have complete knowledge of the callstack. This kind of emergent complexity actually is much better suited to cases where you’ve created your own instance of immer, not used the global one. Because it allows you to guarantee that the assumptions you’ve baked into your code’s behavior can’t be changed out from under you by the caller you don’t control. Which is why it makes sense to flip the configuration. 

Obviously I can convince you if you don’t want to listen. But all I’m saying is the current architecture is a footgun. And by switching the default you’d actually unlock new use cases with no extra work—pretty much the ideal use case when designing a library like this one.

**Comment by mweststrate** [comment]
Would you mind create a PR for that @ianstormtaylor? That would be really helpful to avoid confusion in the future
